File: 0061-rotate-list/0061-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head == NULL || head->next == NULL || k == 0) return head;

        ListNode* temp = head;
        int cnt = 1;

        while(temp->next != NULL){
            temp = temp->next;
            cnt++;
        }

        temp->next = head;

        k = k % cnt;

        int steps = cnt - k;
        ListNode* newTail = head;

        for(int i = 1; i< steps; i++){
            newTail = newTail->next;
        }

        ListNode* newHead = newTail->next;
        newTail->next = NULL;

        return newHead;

    }
};

File: 0061-rotate-list/0061-rotate-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: Optional[ListNode]
        :type k: int
        :rtype: Optional[ListNode]
        """

        if not head or not head.next or k == 0:
            return head

        n = 1
        tail = head
        while tail.next:
            tail = tail.next
            n += 1

        tail.next = head

        k = k % n

        steps = n - k
        newTail = head
        for _ in range(steps - 1):
            newTail = newTail.next

        newHead = newTail.next
        newTail.next = None

        return newHead
        